lib/sshutils/marshal_test.go

```diff
@@ -0,0 +1,67 @@
+/*
+Copyright 2022 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package sshutils
+
+import (
+	"testing"
+
+	"github.com/stretchr/testify/require"
+
+	"github.com/gravitational/teleport/lib/auth/native"
+)
+
+func TestMarshalUnmarshalKnownHosts(t *testing.T) {
+	priv, err := native.GeneratePrivateKey()
+	require.NoError(t, err)
+	authorizedKey := priv.MarshalSSHPublicKey()
+	knownHosts := []KnownHost{
+		{
+			AuthorizedKey: authorizedKey,
+			ProxyHost:     "proxy.example.com",
+			Hostname:      "cluster1",
+			Comment: map[string][]string{
+				"logins": {"root"},
+			},
+		}, {
+			AuthorizedKey: authorizedKey,
+			ProxyHost:     "proxy.example.com",
+			Hostname:      "cluster2",
+		}, {
+			AuthorizedKey: authorizedKey,
+			Hostname:      "cluster3",
+		},
+	}
+
+	var knownHostsFile [][]byte
+	for _, kh := range knownHosts {
+		line, err := MarshalKnownHost(kh)
+		require.NoError(t, err)
+		knownHostsFile = append(knownHostsFile, []byte(line))
+	}
+
+	parsedKnownHosts, err := UnmarshalKnownHosts(knownHostsFile)
```

Comment:
I don't think this type of interoperability is supported or intended. These methods are only used when dealing with Teleport `known_hosts` (`~/.tsh/known_hosts`). If you `UnmarshalKnownHosts` some entry from `~/.ssh/known_hosts` and then `MarshalKnownHost`, it will add the "@cert-authority" marker, the wildcard host name, and the `type=host` comment. Additionally, ssh known hosts may hash the host name and neither we nor `ssh.ParseKnownHosts` handles that case.

To further clarify, `MarshalKnownHost` is only used to create new known host entries for Teleport CA's when logging in or generating an identity file. `UnmarshalKnownHosts` reads those previously written entries.

Are we concerned about handling hosts that users move from `~/.ssh/known_hosts` to `~/.tsh/known_hosts`? Let me know if there is something I am missing.

Reply:
Oh, I'm sorry
I was not aware that this was exclusive to the `.tsh/known_hosts`